### PR TITLE
add unnecessary_self_imports lint

### LINKED ISSUE #6552: A new lint for imports ending in `::{self};`
### What it does

It detects imports that end in `::{self};` and emits a warning for them.

### Categories (optional)

- Kind: Perhaps `clippy::style`.

The advantage of the recommended code over the original code is that it's shorter and simpler.

### Drawbacks

None.

### Example

```rust
use std::io::{self};
```

Could be written as:

```rust
use std::io;
```

### PR DESCRIPTION
**Pull request by ebobrow**
fixes #6552 

changelog: add `unnecessary_self_imports` lint

### CODE REVIEW

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        if_chain! {
+            if !in_macro(item.span);
+            if let ItemKind::Use(path, UseKind::ListStem) = &item.kind;
+            if let Some(ident) = is_self_import(cx, item, path);
+            if let Some(Res::Def(DefKind::Mod, def_id)) = path.segments[path.segments.len() - 2].res;
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:43]: This probably can't panic. But I would use pattern matching here either way.

```suggestion
            if let &[.., module, _self] = &path.segments;
            if let Some(Res::Def(DefKind::Mod, def_id)) = module.res;
```

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            if let Some(last) = path.segments.last();
+            if !mod_contains_item(cx, def_id, last.ident) || !item_is_in_scope(cx, ident, path.span);
+            then {
+                let mut applicability = Applicability::MachineApplicable;
+                let snippet = if ident == path.segments[path.segments.len() - 1].ident {
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:48]: I always get uneasy, when indexing. You already got the last path segment above.

```suggestion
                let snippet = if ident == last.ident {
```

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                    let adjusted_span = path.span.with_hi(path.span.hi() - BytePos(8));
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:49]: `BytePos` should be avoided when constructing spans. I guess you want to get the span of the `::{self}` part of the path? You should be able to do this with `path.span.with_lo(module.ident.span.hi())`.

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        }
+    }
+}
+
+fn is_self_import(cx: &LateContext<'_>, current_import: &Item<'_>, current_import_path: &Path<'_>) -> Option<Ident> {
+    let definitions = cx.tcx.hir().definitions();
+    let mut amt = 0;
+    let mut ident: Option<Ident> = None;
+
+    for def in definitions.iter_local_def_id() {
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:82]: I don't really get the point of this function. Why do you have to check **all** definitions in a `crate`? Couldn't you just check if the last path segment is `self`?

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        }
+    }
+    false
+}
+
+fn item_is_in_scope(cx: &LateContext<'_>, ident: Ident, current_import: Span) -> bool {
+    let definitions = cx.tcx.hir().definitions();
+    let mut outer_mod: Option<Span> = None;
+
+    for def in definitions.iter_local_def_id() {
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:137]: Could you add some comments what `mod_contains_item` and `item_in_scope` do? Iterating through **all** definitions in a crate seems a little overkill to me for this lint.

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                let mut applicability = Applicability::MachineApplicable;
+                let snippet = if ident == path.segments[path.segments.len() - 1].ident {
+                    let adjusted_span = path.span.with_hi(path.span.hi() - BytePos(8));
+                    format!(
+                        "{}",
+                        snippet_with_applicability(cx, adjusted_span, "..", &mut applicability)
+                    )
+                } else {
+                    let adjusted_span = path.span.until(ident.span);
+                    let adjusted_span = adjusted_span.with_hi(adjusted_span.hi() - BytePos(11));
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_self_imports.rs:56]: Same with `BytePos` here.

**Review by flip1995 — COMMENTED** [review]
The functions you implement to check if something is in scope seem really resource intensive to me. Have you looked in `rustc` if there is a query to determine if a item is in scope?

File: clippy_lints/src/unnecessary_self_imports.rs (hunk 4 above)
Comment (ebobrow) [reply clippy_lints/src/unnecessary_self_imports.rs:82]: I could be missing something, but when `self` is included in an import it's not included as a path segment, so the last path segment is whatever `self` refers to. My thinking was to get the last path segment from a `ListStem` item and compare that against the last path segment in `Single` items contained in that scope. If they're the same, then the import includes `self`. I'll look into ways to optimize it though, because I definitely don't need all the defnitions in the crate.

Comment (flip1995) [reply clippy_lints/src/unnecessary_self_imports.rs:82]: IIRC I did something similar in the `wildcard_imports` lint to build the suggestion. You shouldn't have to go over every definition. Let me look that up.

**Comment by flip1995** [comment]
TL;DR: Implement this lint on the AST, not the HIR. It'll be way easier.

I just noticed that you can't really implement this lint on the HIR, since the information if a `self` was involved in an import isn't available there anymore. It just gets lowered to 

```rust
use path::to::module::{}; 
use path::to::module;
```

However on the AST you should be able to determine if it is an import of just `self` and if it has a `as something` pretty easily

<details>
<summary>AST `use` expansion</summary>

```
        Item {
            attrs: [],
            id: NodeId(4294967040),
            span: main.rs:5:1: 5:15 (#0),
            vis: Visibility {
                kind: Inherited,
                span: main.rs:5:1: 5:1 (#0),
                tokens: None,
            },
            ident: #0,
            kind: Use(
                UseTree {
                    prefix: Path {
                        span: main.rs:5:5: 5:6 (#0),
                        segments: [
                            PathSegment {
                                ident: a#0,
                                id: NodeId(4294967040),
                                args: None,
                            },
                        ],
                        tokens: None,
                    },
                    kind: Nested(
                        [
                            (
                                UseTree {
                                    prefix: Path {
                                        span: main.rs:5:9: 5:13 (#0),
                                        segments: [
                                            PathSegment {
                                                ident: self#0,
                                                id: NodeId(4294967040),
                                                args: None,
                                            },
                                        ],
                                        tokens: None,
                                    },
                                    kind: Simple(
                                        None,
                                        NodeId(4294967040),
                                        NodeId(4294967040),
                                    ),
                                    span: main.rs:5:9: 5:13 (#0),
                                },
                                NodeId(4294967040),
                            ),
                        ],
                    ),
                    span: main.rs:5:5: 5:14 (#0),
                },
            ),
            tokens: None,
        },

```
</details>

[`UseTree` Documentation](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.UseTree.html)

You'll then have the problem to determine if the imported module is already there in form of another item (at least I think that is what you try to do with the other 2 functions?). But this isn't really a problem. `rustc` will already error if it is. And if you have an alias, you can just remove the `self` and use `use path::to::module as alias` directly.

**Comment by camsteffen** [comment]
It's a bit more complicated.

* `use a::b::{self}` -> import the module `a::b`
* `use a::b;` -> import the module and/or item `a::b`

So changing to the latter may cause "b is defined multiple times". [Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27510da5100b94441e1507f75b79f1f9)

I think a late pass will be necessary to detect this. And `UseKind::ListStem` can maybe be used to detect `::{self}`? Not sure. Seems hacky.

**Comment by flip1995** [comment]
>     * `use a::b;` -> import the module and/or item `a::b`

Checking another case which shouldn't be a concern for this lint: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fe37f9add3a5a7f9e2b38591a8ef41ef)

So you only have to check which items are loaded in the current module. So instead of checking every `use` item you can just check on module level and walk over all items and lint on `use` items iff they aren't in scope in another form.

**Comment by ebobrow** [comment]
Couldn't `self` still be removed if an item with the same name exists in scope but not in the module? I'm not sure how to check for that in the AST.

**Comment by flip1995** [comment]
Oh yeah, your right. The item could still be in scope from the surrounding scope... I'm starting to think that there is no feasible way to implement this. Your method works, but I hope that you agree that going over all defined items multiple times is too performance heavy.

Determining what would happen if we change code without recompiling is just really hard in general. 

I see two ways going forward:
1. Not adding this lint. I'm not sure how helpful this lint is in the first place. When someone uses `::{self}` it is either intentional or just a refactor artifact. It doesn't hurt to have it. An expensive lint for code style isn't justified IMO.
2. Ignore that the item could already be in scope and implement it as a restriction lint with a statement in the documentation that this lint isn't meant to be permanently enabled, but only to quickly check, e.g. after refactoring, if there are artifacts like this in the code.

**Comment by camsteffen** [comment]
> if an item with the same name exists in scope but not in the module

I'm not following...

**Comment by ebobrow** [comment]
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e9e0c125dd34a7a682e9b00676d8bdaa)

I think @flip1995 is right. On the HIR, this lint is too resource intensive and it might not be possible on the AST. I don't know if it's worthwhile to have.

**Comment by camsteffen** [comment]
Hrrrmmm. Yeeahh. mrr

**Comment by flip1995** [comment]
> > if an item with the same name exists in scope but not in the module
> 
> I'm not following...

Or maybe not... [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=577acaf2476bd293d1e439ab01c1e5bb) 

Anyway, this is way too scuffed to figure out what works and what doesn't when the `::{self}` might get removed to make it FP free.

**Comment by camsteffen** [comment]
Yes too scuffed! I can imagine the restriction lint begin useful as in "if removing `::{self}` breaks something, you're doing it wrong."

**Comment by ebobrow** [comment]
Good point. I'll change this to a restriction lint and put a warning in the documentation.

**Comment by camsteffen** [comment]
I think we should also

```
diag.span_note("this will change semantics; any non-module items at the same path will also be imported");
```

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::kw::SelfLower;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for imports ending in `::{self};`.
+    ///
+    /// **Why is this bad?** In most cases, this can be written much more cleanly by omitting `self`.
+    ///
+    /// **Known problems:** Sometimes removing `::{self}` will break code (https://github.com/rust-lang/rustfmt/issues/3568).
+    /// This lint should not be permanently enabled.
```
Comment (camsteffen) [inline clippy_lints/src/unnecessary_self_imports.rs:16]: It's not necessarily wrong to enable the lint for a project.

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            if let PathSegment { ident, .. } = prefix.segments[0];
+            if ident.name == SelfLower;
+
+            then {
+                let adjusted_span = item.span.with_hi(span.hi());
+                let snippet = if let UseTreeKind::Simple(Some(alias), ..) = kind {
+                    format!(
+                        "{} as {};",
+                        snippet(cx, adjusted_span, ".."),
+                        snippet(cx, alias.span, "..")
```
Comment (camsteffen) [inline clippy_lints/src/unnecessary_self_imports.rs:50]: I don't think any `snippet` calls are needed.

```suggestion
                        use_tree.ident(),
                        alias,
```

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl EarlyLintPass for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if_chain! {
+            if let ItemKind::Use(
+                UseTree { kind: UseTreeKind::Nested(nodes), prefix: Path { span, .. }, .. }
+            ) = &item.kind;
```
Comment (camsteffen) [inline clippy_lints/src/unnecessary_self_imports.rs:38]: Just making `use_tree` available for a later suggestion...

```suggestion
            if let ItemKind::Use(use_tree) = &item.kind;
```

File: tests/ui/unnecessary_self_imports.rs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// run-rustfix
+#![warn(clippy::unnecessary_self_imports)]
+#![allow(unused_imports, dead_code)]
+
+use std::collections::hash_map::{self, *};
+use std::fs::{self as alias};
+use std::io::{self};
```
Comment (camsteffen) [inline tests/ui/unnecessary_self_imports.rs:7]: One more test with `::{self, other_thing}` would be good.

**Review by camsteffen — CHANGES_REQUESTED** [review]
Just some nits.

File: clippy_lints/src/unnecessary_self_imports.rs (hunk 8 above)
Comment (ebobrow) [reply clippy_lints/src/unnecessary_self_imports.rs:50]: When I add `use_tree.ident()`, the program panics:
```
thread 'rustc' panicked at '`UseTree::ident` can only be used on a simple import', compiler/rustc_ast/src/ast.rs:2345:18
```
I can remove the `snippet` call for the alias, but it looks like I have to keep it in for the path.

Comment (camsteffen) [reply clippy_lints/src/unnecessary_self_imports.rs:50]: How about `use_tree.prefix.segments.last()?.ident`?

Comment (ebobrow) [reply clippy_lints/src/unnecessary_self_imports.rs:50]: That works! I have to make the span a little wonky so that it only covers the last segment in the path plus the `::{self}`, but I'm assuming that's alright?

File: clippy_lints/src/unnecessary_self_imports.rs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl EarlyLintPass for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if_chain! {
+            if let ItemKind::Use(use_tree) = &item.kind;
+            if let UseTree { kind: UseTreeKind::Nested(nodes), .. } = use_tree;
+            if let [(UseTree { prefix, kind, .. }, _)] = &**nodes;
+            if let [PathSegment { ident, .. }, ..] = &*prefix.segments;
+            if ident.name == kw::SelfLower;
```
Comment (camsteffen) [inline clippy_lints/src/unnecessary_self_imports.rs:41]: I would opt for some smaller patterns and fewer field bindings. But it's up to you.

```suggestion
            if let UseTreeKind::Nested(nodes) = use_tree.kind;
            if let [(self_tree, _)] = &**nodes;
            if let [self_seg] = &*self_tree.prefix.segments;
            if self_seg.ident.name == kw::SelfLower;
```

**Review by camsteffen — APPROVED** [review]
One more minor suggestion that you can take or leave. Otherwise looks good!

**Comment by ebobrow** [comment]
Thanks! I like that.

**Review by flip1995 — APPROVED** [review]
LGTM now. @camsteffen if you're also ok with this PR, please go ahead and merge it.

@ebobrow Thanks for all your work on it, even though you had to delete most of it again. But now we learned what information about imports are available where, which will help us advising on issues about imports in the future,, which is definitely a win.
